### PR TITLE
Restoring init container support.

### LINKED ISSUE #2139: Support custom init containers and volumes in KServe
/kind feature

**Describe the solution you'd like**

The current KServe don't support init containers and uses webhook to inject storage initializer due to the limitation in KNative previously. [Ref](https://github.com/kserve/kserve/blob/v0.8.0/pkg/webhook/admission/pod/storage_initializer_injector.go#L82)

KNative has support [init containers](https://knative.dev/v1.2-docs/serving/configuration/feature-flags/#kubernetes-init-containers) and [PVC](https://knative.dev/v1.2-docs/serving/configuration/feature-flags/#kubernetes-persistentvolumeclaim-pvc) in `v1.2` natively.

Is it possible to support custom `init containers` and `pvc` in KServe as well so that users can use their own storage initializer and PVC, which provide more flexibility to the users. Thanks.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

### PR DESCRIPTION
**Pull request by rachitchauhan43**
Signed-off-by: rachitchauhan43 <rachitchauhan43@gmail.com>

**What this PR does / why we need it**:
Use cases are: 
1. In case of Inference graphs, we need to make calls to external services from within a pre/post-process method and for that we need to construct auth headers using secrets loaded by init-container.
2. Similarly, there is a use case where have sensitive data that needs to be decrypted first inside predict method using secret loaded using init container.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2139

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevent result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by rachitchauhan43** [comment]
## How I tested ?

Used this yaml for ISVC:

```apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: test-init-container
  namespace: kserve-test
spec:
  predictor:
    volumes:
      - emptyDir: { }
        name: shared-data
    initContainers:
      - image: docker.myorg.com/data-mlplatform/utils/secret-loader:latest
        name: secrets-loader
        volumeMounts:
          - mountPath: /secrets
            name: shared-data
        command: [ 'sh',
                   '-c',
                   'date > /secrets/my-value' ]
    containers:
      - args:
          - --model_name
          - qbo-in
        image: docker.myorg.com/personal/myuser/qbo-in:v1
        name: kserve-container
        volumeMounts:
          - mountPath: /secrets
            name: shared-data
        resources:
          requests:
            cpu: 101m
            memory: 256Mi
```

As you can see below, i ssh into container and can see my secrets mounted:
Screen Shot 2022-10-19 at 9 09 10 AM


See the finally YAML with status and spec both:

[isvc_with_initcontainer_yaml.txt](https://github.com/kserve/kserve/files/9822323/isvc_with_initcontainer_yaml.txt)

File: Makefile [hunk 1]
```diff
@@ -153,7 +153,7 @@ manifests: controller-gen
 	yq d -i config/crd/serving.kserve.io_inferenceservices.yaml 'spec.versions[0].schema.openAPIV3Schema.properties.spec.properties.*.properties.*.required'
 	#remove ephemeralContainers properties for compress crd size https://github.com/kubeflow/kfserving/pull/1141#issuecomment-714170602
 	yq d -i config/crd/serving.kserve.io_inferenceservices.yaml 'spec.versions[0].schema.openAPIV3Schema.properties.spec.properties.*.properties.ephemeralContainers'
-	yq d -i config/crd/serving.kserve.io_inferenceservices.yaml 'spec.versions[0].schema.openAPIV3Schema.properties.spec.properties.*.properties.initContainers'
+	#yq d -i config/crd/serving.kserve.io_inferenceservices.yaml 'spec.versions[0].schema.openAPIV3Schema.properties.spec.properties.*.properties.initContainers'
```
Comment (yuzisun) [inline Makefile]: Did you regenerate the crd? I think the crd schema needs to be updated

Comment (rachitchauhan43) [reply Makefile]: @yuzisun : Added the updated crds too.

**Comment by yuzisun** [comment]
Thanks @rachitchauhan43 ! I will re-generate the swagger json later and updated the doc for the process of code-gen.
/lgtm
/approve

**Comment by rachitchauhan43** [comment]
> Thanks @rachitchauhan43 ! I will re-generate the swagger json later and updated the doc for the process of code-gen. /lgtm /approve

@yuzisun: Can you point me where you updated the doc for `the process of code-gen` ?

**Comment by yuzisun** [comment]
> > Thanks @rachitchauhan43 ! I will re-generate the swagger json later and updated the doc for the process of code-gen. /lgtm /approve
> 
> @yuzisun: Can you point me where you updated the doc for `the process of code-gen` ?

It is here in the developer guide https://kserve.github.io/website/master/developer/developer/#iterating

**Pull request merged**